packages/core/test/final.test.ts

```diff
@@ -574,4 +574,206 @@ describe('final states', () => {
+    expect(spy.mock.calls).toMatchInlineSnapshot(`
+      [
+        [
+          {
+            "output": undefined,
+            "type": "xstate.done.state.(machine)",
+          },
+        ],
+      ]
+    `);
+  });
+
+  it('root output should be called with a "xstate.done.state.*" event of the parallel root when a final descendant is reached 2 parallel levels deep', () => {
+    const spy = jest.fn();
+    const machine = createMachine({
+      type: 'parallel',
+      states: {
+        a: {
+          type: 'parallel',
+          states: {
+            b: {
+              initial: 'c',
+              states: {
+                c: {
+                  type: 'final'
+                }
+              }
+            }
+          }
+        }
+      },
+      output: ({ event }) => {
+        spy(event);
+      }
+    });
+
+    createActor(machine).start();
+
+    expect(spy.mock.calls).toMatchInlineSnapshot(`
+      [
+        [
+          {
+            "output": undefined,
+            "type": "xstate.done.state.(machine)",
+          },
+        ],
+      ]
+    `);
+  });
+
+  it('onDone of an outer parallel state should be called with its own "xstate.done.state.*" event when its direct parallel child completes', () => {
+    const spy = jest.fn();
+    const machine = createMachine({
+      initial: 'a',
+      states: {
+        a: {
+          type: 'parallel',
+          states: {
+            b: {
+              type: 'parallel',
+              states: {
+                c: {
+                  initial: 'd',
+                  states: {
+                    d: {
+                      type: 'final'
+                    }
+                  }
+                }
+              }
+            }
+          },
+          onDone: {
+            actions: ({ event }) => {
+              spy(event);
+            }
+          }
+        }
+      }
+    });
+    createActor(machine).start();
+
+    expect(spy.mock.calls).toMatchInlineSnapshot(`
+      [
+        [
+          {
+            "output": undefined,
+            "type": "xstate.done.state.(machine).a",
+          },
+        ],
+      ]
+    `);
+  });
+
+  it('onDone should not be called when the machine reaches its final state', () => {
+    const spy = jest.fn();
+    const machine = createMachine({
+      type: 'parallel',
+      states: {
+        a: {
+          type: 'parallel',
+          states: {
+            b: {
+              initial: 'c',
+              states: {
+                c: {
+                  type: 'final'
+                }
+              },
+              onDone: {
+                actions: spy
+              }
+            }
+          },
+          onDone: {
+            actions: spy
+          }
+        }
+      },
+      onDone: {
+        actions: spy
+      }
+    });
+    createActor(machine).start();
+
+    expect(spy).not.toHaveBeenCalled();
+  });
+
+  it('machine should not complete when a parallel child of a compound state completes', () => {
```

Comment:
Imagine that you have `a.onDone: 'other'` here. With it both would be true:
- we should complete the machine
- we should go to `'other'`

Overall, `onDone: 'something'` is a little bit wonky at times since we push those events to the internal queue but break the main event loop when we reach the final state. So it's always possible that those events from the loop would just get discarded/ignored as we stop processing the leftover events after reaching a final state. 

This scenario looks very much OK though and I think it's better for this `onDone` to be executed instead of putting the whole machine in its done state. A change to this would be a further deviation from the SCXML spec as well